src/main/java/org/xbill/DNS/TSIG.java

```diff
@@ -749,15 +849,16 @@ public int verify(Message m, byte[] b) {
 
       nresponses++;
       if (nresponses == 1) {
-        int result = key.verify(m, b, lastTSIG);
-        lastTSIG = tsig;
+        int result = key.verify(m, b, queryTsig, true, sharedHmac);
+        hmacAddSignature(sharedHmac, tsig);
+        lastsigned = nresponses;
         return result;
       }
 
       if (tsig != null) {
-        int result = key.verify(m, b, lastTSIG, false);
+        int result = key.verify(m, b, null, false, sharedHmac);
```

Comment:
Very clever.  I think this satisfies all my concerns about backward compatibility.